### PR TITLE
Issue 930: L044 -- Max Recursion Depth

### LINKED ISSUE #930: L044: Max Recursion Depth
Came across this `CRITICAL` error when linting a few models in our dbt project using sqlfluff `0.5.0`:

#### Steps to reproduce:
```
$ sqlfluff lint models/sales/hubspot_engagement_calls.sql
CRITICAL   [L044] Applying rule L044 threw an Exception: maximum recursion depth exceeded while calling a Python object 
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/base.py", line 290, in crawl
    res = self._eval(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/std/L044.py", line 138, in _eval
    return self._analyze_result_columns(select_info, dialect, queries)
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/std/L044.py", line 106, in _analyze_result_columns
    self._analyze_result_columns(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/std/L044.py", line 106, in _analyze_result_columns
    self._analyze_result_columns(
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/std/L044.py", line 106, in _analyze_result_columns
    self._analyze_result_columns(
  [Previous line repeated 976 more times]
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/std/L044.py", line 90, in _analyze_result_columns
    for wildcard in select_info.get_wildcard_info():
  File "/usr/local/lib/python3.8/site-packages/sqlfluff/core/rules/analysis/select_crawler.py", line 119, in get_wildcard_info
    WildcardInfo(
  File "<string>", line 1, in __new__
RecursionError: maximum recursion depth exceeded while calling a Python object
```

#### SQLFluff Verbose Output
[sqlfluff_verbose_out.txt](https://github.com/sqlfluff/sqlfluff/files/6268014/sqlfluff_verbose_out.txt)

#### Version

```sh
$ sqlfluff --version
sqlfluff, version 0.5.0
```

#### Configuration
```
[sqlfluff]
dialect = snowflake
templater = dbt
# dbt templating does not keep trailing new lines (L009)
exclude_rules = L007,L009,L031,L033,L034

[sqlfluff:rules]
max_line_length = 120
comma_style = trailing

[sqlfluff:rules:L010]
capitalisation_policy = upper
```

#### SQL

```sql
{{
    config(
        materialized = 'table'
    )
}}

with
    hubspot__engagement_calls as (
        select * from {{ ref('hubspot__engagement_calls') }}
    )

select * from hubspot__engagement_calls
```

#### Compiled SQL

```sql

with
    hubspot__engagement_calls as (
        select * from ANALYTICS_DWALLACE.PUBLIC_intermediate.hubspot__engagement_calls
    )

select * from hubspot__engagement_calls
```

### PR DESCRIPTION
**Pull request by barrywhart**
Resolves #930 

The coverage report is failing, but only because I added code to detect when linter tests cause unexpected exceptions -- I think the PR is ready for review and potential merge, unless we feel strongly about adding tests for the test framework.

### CODE REVIEW

File: src/sqlfluff/testing/rules.py [hunk 1]
```diff
@@ -59,6 +59,9 @@ def assert_rule_fail_in_sql(code, sql, configs=None, line_numbers=None):
     linted = Linter(config=cfg).lint_string(sql, fix=True)
     lerrs = linted.get_violations()
     print("Errors Found: {0}".format(lerrs))
+    for e in lerrs:
```
Comment (barrywhart) [inline src/sqlfluff/testing/rules.py:62]: In reproducing the bug and testing the fix, I noticed that the tests were "eating" the infinite recursion error, i.e. if a test case had a `fail_str`, the infinite recursion was caught and treated as an expected lint error. With this change, any **unexpected** exceptions will automatically cause the test to fail. I think this is the behavior we want.

File: src/sqlfluff/testing/rules.py [hunk 2]
```diff
@@ -59,6 +59,9 @@ def assert_rule_fail_in_sql(code, sql, configs=None, line_numbers=None):
...
+        if e.desc().startswith("Unexpected exception"):
+            pytest.fail(f"Linter failed with {e.desc()}")
```
Comment (barrywhart) [inline src/sqlfluff/testing/rules.py:64]: This is the line causing the coverage check to complain -- if we were hitting this line, that'd mean we have a failing test. I think having this line uncovered is fine.

File: src/sqlfluff/core/rules/analysis/select_crawler.py [hunk 3]
```diff
@@ -120,6 +120,8 @@ def get_wildcard_info(self) -> List[WildcardInfo]:
                             seg,
                             [
                                 alias_info.ref_str
+                                if alias_info.aliased
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/analysis/select_crawler.py:123]: This fixes the real issue: The rule was confusing two things:
* A table whose base name was "blah"
* A CTE named "blah"

File: src/sqlfluff/core/rules/std/L044.py [hunk 4]
```diff
@@ -104,7 +104,7 @@ def _analyze_result_columns(
                             if wildcard_table in queries:
                                 # Wildcard refers to a CTE. Analyze it.
                                 self._analyze_result_columns(
-                                    queries[wildcard_table], dialect, queries
```
Comment (barrywhart) [inline src/sqlfluff/core/rules/std/L044.py:107]: This fixes another issue I discovered in tracking down the issue: A CTE that queries itself was also causing infinite recursion.

Comment (alanmcruickshank) [reply src/sqlfluff/core/rules/std/L044.py:107]: NIce 👍 

Initially I didn't get this, but I think it's actually quite an elegant solution.

File: src/sqlfluff/testing/rules.py (hunk 1 above)
Comment (alanmcruickshank) [reply src/sqlfluff/testing/rules.py:62]: Good catch

**Review by alanmcruickshank — APPROVED** [review]
Nice.

**Pull request merged**